src/fabtrimino.c

```diff
@@ -0,0 +1,250 @@
+        fill_Z_squares(matrix);
+        break;
+    case S:
+        fill_S_squares(matrix);
+        break;
+    case O:
+        fill_O_squares(matrix);
+        break;
+    default:
+        printf("INVALID SHAPE: %d\n", shape);
+        break;
+    }
+}
+
+struct fabtrimino* fab_make(enum shape shape)
+{
+    struct fabtrimino* fab = calloc(1, sizeof(*fab));
+
+    fab->matrix = calloc(4, sizeof(struct square**));
+    for (int i = 0; i<4; i++)
+    {
+        fab->matrix[i] = calloc(4, sizeof(struct square*));
+    }
+    nullify_squares_matrix(fab->matrix, 4);
+
+    fill_shape_squares(fab->matrix, shape);
+
+    fab->shape = shape;
+    fab->pos = NULL;
+
+    return fab;
+}
+
+void free_square(struct square* square)
+{
+    free(square);
+}
+
+void fab_free(struct fabtrimino* fab)
+{
+    for (int i = 0; i < 4; i++) // TODO: change the 4 to matrix.size once the matrix is a struct
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            if (fab->matrix[i][j] != NULL)
+            {
+                free_square(fab->matrix[i][j]);
+            }
+        }
+        free(fab->matrix[i]);
+    }
+    free(fab->matrix);
+    if (fab->pos != NULL)
+    {
+        free(fab->pos);
+    }
+    free(fab);
+}
+
+struct color {
+    Uint8 r;
+    Uint8 g;
+    Uint8 b;
+};
+
+struct color shape_color(enum shape shape)
+{
+    struct color col;
+    switch(shape) {
+    case T:
+        col.r = 0x80;
+        col.g = 0x00;
+        col.b = 0x80;
+        return col;
+    case I:
+        col.r = 0x00;
+        col.g = 0xFF;
+        col.b = 0xFF;
+        return col;
+    case L:
+        col.r = 0xFF;
+        col.g = 0xA5;
+        col.b = 0x00;
+        return col;
+    case J:
+        col.r = 0x00;
+        col.g = 0x00;
+        col.b = 0xFF;
+        return col;
+    case S:
+        col.r = 0x00;
+        col.g = 0x80;
+        col.b = 0x00;
+        return col;
+    case Z:
+        col.r = 0xFF;
+        col.g = 0x00;
+        col.b = 0x00;
+        return col;
+    case O:
+        col.r = 0xFF;
+        col.g = 0xFF;
+        col.b = 0x00;
+        return col;
+    default:
+        col.r = 0xAA;
+        col.g = 0xBB;
+        col.b = 0xCC;
+        return col;
+    }
+}
+
+void draw_square(struct square* square, struct vector* position, SDL_Renderer* renderer)
+{
+    struct color color = shape_color(square->shape);
+
+    SDL_SetRenderDrawColor(renderer, color.r, color.g, color.b, 0xFF);
+
+    SDL_Rect s = {
+        .x = position->x * GRID_SQUARE_LENGTH,
+        .y = position->y * GRID_SQUARE_LENGTH,
+        .w = GRID_SQUARE_LENGTH,
+        .h = GRID_SQUARE_LENGTH
+    };
+    SDL_RenderFillRect(renderer, &s);
+}
+
+void fab_draw(struct fabtrimino* fab, SDL_Renderer* renderer, struct vector* offset)
+{
+    for (int i = 0; i < 4; i++)
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            if (fab->matrix[i][j] != NULL)
+            {
+                struct vector* position = calloc(1, sizeof(*position));
```

Comment:
You really don't have to heap allocate this structure which leads me to believe that you might not know the distinction between heap allocation and stack/static allocations.

Essentially, you create any type with a size like so:
```
struct vector position = {0};
```
when you do this, the vector will be allocated on the stack which essentially just involves incrementing the stack pointer. This also has the benefit that the struct will be reclaimed when the function/scope exits without having to call free at all. 

Given that stack allocated structures are far less error prone and efficient it should be your default and you should need a good reason to go to the heap.